lib/nbd.c

```diff
@@ -0,0 +1,792 @@
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#ifndef CURL_DISABLE_NBD
+
+/* Curl support for the NBD protocol.
+ *
+ * Only simple uploads and downloads of the whole image are supported.
+ * Only servers using the "fixed newstyle" protocol are supported.
+ *
+ * Some other missing features:
+ *
+ * - Connect over Unix domain socket.
+ *
+ * - TLS authentication and encryption.
+ *
+ * - Detect if upload block is all zeroes and use NBD_CMD_WRITE_ZEROS.
+ *
+ * - Range requests (upload/download part of the image).
+ *
+ * - Support for: trim, write zeroes, flush.
+ *
+ * - Setting FUA flag in requests.
+ *
+ * - Support for server flags: readonly, rotational.
+ *
+ * - List export names.
+ */
+
+#ifdef HAVE_NETINET_IN_H
+#include <netinet/in.h>
+#endif
+#ifdef HAVE_ARPA_INET_H
+#include <arpa/inet.h>
+#endif
+#ifdef HAVE_UTSNAME_H
+#include <sys/utsname.h>
+#endif
+#ifdef HAVE_NETDB_H
+#include <netdb.h>
+#endif
+#ifdef __VMS
+#include <in.h>
+#include <inet.h>
+#endif
+
+#if (defined(NETWARE) && defined(__NOVELL_LIBC__))
+#undef in_addr_t
+#define in_addr_t unsigned long
+#endif
+
+#include <curl/curl.h>
+#include "urldata.h"
+#include "connect.h"
+#include "progress.h"
+#include "sendf.h"
+#include "multiif.h"
+#include "transfer.h"
+#include "escape.h"
+#include "nbd.h"
+/* The last #include file should be: */
+#include "memdebug.h"
+
+/* Maximum we will read from the socket, which is also the size of
+ * the internal recv_buf buffer that we allocate per handle.
+ */
+#define NBD_MAX_MESSAGE_SIZE 8192
+
+/* The maximum size of a request we will make to the server.
+ *
+ * nbdkit supports up to 64MB per request, but other servers may
+ * support less than this so choose a smaller number than the
+ * theoretical maximum.
+ */
+#define NBD_MAX_REQUEST_SIZE (4*1024*1024)
+
+#if defined(_MSC_VER) || defined(__ILEC400__)
+#  define PACK
+#  pragma pack(push)
+#  pragma pack(1)
+#elif defined(__GNUC__)
+#  define PACK __attribute__((packed))
+#else
+#  define PACK
+#endif
+
+struct handshake {
+  char nbdmagic[8];             /* "NBDMAGIC" */
+  char version[8];              /* "IHAVEOPT"  */
+  unsigned short gflags;        /* global flags */
+#define NBD_FLAG_FIXED_NEWSTYLE 1
+#define NBD_FLAG_NO_ZEROES      2
+} PACK;
+
+struct option {
+  char version[8];              /* "IHAVEOPT"  */
+  unsigned int option;          /* NBD_OPT_* */
+#define NBD_OPT_EXPORT_NAME 1
+  unsigned int len;             /* Length of option data. */
+  char data[1];                 /* Option data. */
+} PACK;
+
+struct export_data {
+  curl_off_t size;              /* Size of the export. */
```

Comment:
You use `curl_off_t` in a packed struct, which tells me you assume a fixed size here. That's not something `curl_off_t` guarantees, so if this code only works with a 64 bit type, you need to protect the entire code with an #ifdef for that. (smb.c does that)